### ISSUE TITLE
Connection Migration API

### ISSUE THREAD

**Issue by marten-seemann** [quic-go/quic-go#3990]
This is a proposal for adding a Connection Migration API.

The `Config` gains a new field that controls if _the peer_ (not the endpoint itself) is allowed to initiate connection migration:
```go
type Config struct {
    // …
    // AllowConnectionMigration controls if the peer is allowed to initiate connection migration.
    AllowConnectionMigration bool
}
```

The state can be checked on the `ConnectionState`. See below for more explanation on `Path`.
```go
type ConnectionState struct {
     // …
     // ConnectionMigrationDisabled says if it’s possible to migrate this connection to a new interface.
     // The peer has the option to disable migration by sending the disable_active_migration transport parameter.
     ConnectionMigrationDisabled bool
     // Path is the network path that’s currently in use.
     Path *Path
}
```

On the client side, migration is initiated by calling `ProbePath` on the `Connection`:
```go
// PathStatus is the status of a path.
type PathStatus int

const (
     // PathStatusProbing means that the path is being probed (i.e. a PATH_CHALLENGE frame has been sent).
     PathStatusProbing PathStatus = iota
     // PathStatusTimeout means that path probing ran into a timeout,
     // or that a previously successfully probed path was abandoned.
     PathStatusTimeout
     // PathStatusProbeSuccess means that path probing succeeded. It is now possible to switch to this path.
     PathStatusProbeSuccess
     // PathStatusActive means that this is the path that’s used to send QUIC packets
     PathStatusActive
)

// Path is a network path.
type Path struct {
     // Status is the status of this path.
     Status PathStatus
     // Notify is a channel (cap 1) that a new value is added to every time the status changes.
     // Note that the status can change even after path validation succeeded (e.g. because a path times out).
     Notify <-chan struct{}
     // might add some path characteristics (RTT, MTU, loss rate, etc.) here later
}

// Commit switches the QUIC connection to this path.
// It immediately stops sending on the old path, and sends on this new path.
// It’s only valid to call Use if the status is ProbeStatusProbeSuccess.
func (*Path) Commit() error

// Abandon abandons a path.
// It is not possible to abandon the path that’s currently active.
// This stops the sending of keep-alive packets, eventually leading to a timeout of the path.
func (*Path) Abandon() error

// ProbePath probes a new path.
// This triggers send a packet containing a PATH_CHALLENGE frame and receiving a PATH_RESPONSE frame.
// Therefore, probing takes at least one RTT.
// Note that the RTT of the new path might be different from the old path’s RTT.
// It is only valid
// 1. on the client side, and
// 2. if the peer hasn’t disabled connection migration.
// Additionally, this function will return an error if there’s no spare Connection ID available.
// Once successfully probed, the path will be kept alive by regularly sending keep-alive packets.
func (*Connection) ProbePath(*Transport) (Path, error)
```

Why does `ProbePath` take a `Transport`, and not an IP address or a `net.PacketConn`? This is because we need to be able to receive packets sent on the new path, i.e. we need to demultiplex them from other packets that might be received on the same interface. This is _exactly_ what a `Transport` does.
Internally, `ProbePath` adds _all_ the connection IDs that we issued to the peer to the new `Transport` as well. Since in RFC 9000, connection IDs are not bound to paths, we can’t know which CID the peer will use on the new path. As a slight optimization, we might be able to omit CIDs that are already used on other paths.

**Comment by lbsystem** [comment]
come on  I love this

**Comment by yinbao2000** [comment]
In the proposed API, does the client must explicitly initiate `PathProbe` and then `Commit`?  If that is the case, then the connection migration is not transparent to the application. It will also come with performance cost because it's driven by how well the application is developed to make use of the feature. How about adding an option in API where client can 1) specify the path options 2) set the KPIs of migration (similar to the characteristic in `notify`. Example: switch to another path when active path pkt loss is over 20%)?  These settings are done ahead of time and the migration happens later without involvement of the application. I know that this may turn connection migration into a traffic engineering middleware that also have to do the jobs of preventing pingpong of paths. But leaving that job to the application client is probably worse?

**Comment by marten-seemann** [comment]
> If that is the case, then the connection migration is not transparent to the application. 

That is **explicitly** not the intention. Connection migration is an advanced feature, and will require fine tuning depending on the use case.

**Comment by stv0g** [comment]
> That is explicitly not the intention. Connection migration is an advanced feature, and will require fine tuning depending on the use case.

I agree. We need this to be explicit in order to interoperate NAT-traversal / ICE.

There are several proposed RFCs which are somewhat based on QUICs connection migrations:

* https://datatracker.ietf.org/doc/html/draft-perkins-quic-p2p-mux-00
* https://datatracker.ietf.org/doc/html/draft-thatcher-p2p-quic-00
* https://datatracker.ietf.org/doc/html/draft-seemann-quic-nat-traversal-02

**Comment by fortuna** [comment]
Thanks @marten-seemann for advancing this!

When designing an API, I find it helpful to not only describe the APIs, but also to show how they would be used in some common cases, to validate it.

For instance how would an app that uses an `http.Client` that is capable of migrating connections look like?

From the app perspective, the logic can be in the http3.Transport. But how would that work?

I'm trying to figure that out, but here is my understanding of what needs to be done on network change:

1. Create a new connection on the new network
   1. Create a new `net.PacketConn` on the new network.
   1. Create a new `quic.Transport` with the new packet conn.
   1. Call `Transport.DialEarly` to get a new `quic.EarlyConection`
1. Call `Connection.ProbePath` on the new early connection
1. If successful, replace the early connection that the HTTP client is using somehow. We need to do this for all active connections, we will need to intercept and track all HTTP client connections somehow.

It may be helpful to have an Early Connection type that supports replacing the underlying `EarlyConnection`. Or can we just replace the `PacketConn`?

We could also consider adding a `quic.Migrate(quic.Transport, net.PacketConn)`

Does that make sense? i'm still new to QUIC and trying to figure it out.

**Comment by fortuna** [comment]
I'm a bit confused. Is this proposal for the client or server API?

I'm reading the RFC, and according to [Initiating Connection Migration](https://datatracker.ietf.org/doc/html/rfc9000#name-initiating-connection-migra), the client initiates the migration by simply sending packets from the new location.

That means we could simply replace the `net.PacketConn` that the `quic.Transport` and `quic.Connection` are using.
No need for probing, since that's the server's responsibility. Of course, the client code needs to still respond to the probes.

In that case, would it make sense to simply have a `quic.Transport.Migrate(net.PacketConn)` API?

**Comment by marten-seemann** [comment]
> It may be helpful to have an Early Connection type that supports replacing the underlying `EarlyConnection`.

I don't understand. The whole point of connection migration is to keep the connection state.



> I'm reading the RFC, and according to [Initiating Connection Migration](https://datatracker.ietf.org/doc/html/rfc9000#name-initiating-connection-migra), the client initiates the migration by simply sending packets from the new location.
>
> That means we could simply replace the net.PacketConn that the quic.Transport and quic.Connection are using.
No need for probing, since that's the server's responsibility. Of course, the client code needs to still respond to the probes.

This is not correct. The client has to probe the new path before using it. Even if it was not mandated by the RFC, it wouldn't be a smart move to give up a working path for a new path that might or might not work at all.

> For instance how would an app that uses an `http.Client` that is capable of migrating connections look like?

That's a separate question. HTTP/3 is an application and the code is just part of this repo for reasons of convenience (see #4674). This issue is dealing with a transport-layer API. If / how this we'll expose this at the HTTP/3 layer remains to be seen.
